File: app/src/main/java/me/zipi/navitotesla/model/Poi.java
package me.zipi.navitotesla.model;

import java.util.Locale;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@ToString(of = "poiName")
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Data
public class Poi {
    String poiName;
    @Getter(AccessLevel.NONE)
    String roadAddress;
    @Getter(AccessLevel.NONE)
    String address;
    String longitude;
    String latitude;

    public boolean isAddressEmpty() {
        return (roadAddress == null || roadAddress.equals("")) && (address == null || address.equals(""));
    }

    public String getRoadAddress() {
        // roadAddress, address, gps
        if (roadAddress != null && roadAddress.length() > 0) {
            return roadAddress;
        } else {
            return getAddress();
        }
    }

    public String getAddress() {
        if (address != null && address.length() > 0) {
            return address;
        } else {
            return getGpsAddress();
        }
    }

    public String getGpsAddress() {
        return String.format(Locale.getDefault(), "%s,%s", latitude, longitude);
    }

}

File: app/src/main/kotlin/me/zipi/navitotesla/model/Poi.kt
package me.zipi.navitotesla.model

data class Poi(
    val poiName: String? = null,
    private val roadAddress: String? = null,
    private val address: String? = null,
    val longitude: String? = null,
    val latitude: String? = null,
    val packageName: String = "",
    val isDuplicate: Boolean = false,
    /** registered favorite 여부. true 면 roadAddress 컬럼 값 그대로 share (모드 전환 없음). */
    val isFavorite: Boolean = false,
) {
    @Suppress("unused")
    fun isAddressEmpty(): Boolean = roadAddress.isNullOrEmpty() && address.isNullOrEmpty()

    @Suppress("unused")
    fun isGpsEmpty(): Boolean = (longitude == null || latitude == null)

    /**
     * roadAddress 가 "lat,lng" 좌표 문자열인지 검사.
     * favorite 의 GPS 모드 마이그레이션 결과 또는 사용자가 직접 좌표 입력한 경우 매칭.
     * "null,null" (Poi.getGpsAddress 의 null lat/lng fallback) 같은 비숫자 문자열은 제외.
     */
    fun isCoordsAddress(): Boolean {
        val road = roadAddress?.trim() ?: return false
        return COORDS_PATTERN.matches(road)
    }

    companion object {
        private val COORDS_PATTERN = Regex("^-?\\d+(?:\\.\\d+)?,\\s*-?\\d+(?:\\.\\d+)?$")
    }

    fun getRoadAddress(): String {
        // roadAddress, address, gps
        return if (!roadAddress.isNullOrEmpty()) {
            roadAddress
        } else {
            getAddress()
        }
    }

    fun getAddress(): String =
        if (!address.isNullOrEmpty()) {
            address
        } else {
            getGpsAddress()
        }

    fun getGpsAddress(): String = "$latitude,$longitude"
}
